src/Select.tsx

```diff
@@ -1,112 +1,160 @@
 "use client";
 
-import React, { memo, forwardRef, ReactNode, useId, type CSSProperties } from "react";
+import React, { memo, forwardRef, ReactNode, useId, type CSSProperties, ForwardedRef } from "react";
 import { symToStr } from "tsafe/symToStr";
 import { assert } from "tsafe/assert";
 import type { Equals } from "tsafe";
 import { fr } from "./fr";
 import { cx } from "./tools/cx";
 
-export type SelectProps = {
+export type SelectProps<Options extends GenericOption<DefaultOptionValue>[]> = {
+    options: Options;
     className?: string;
     label: ReactNode;
     hint?: ReactNode;
-    nativeSelectProps: React.DetailedHTMLProps<
-        React.SelectHTMLAttributes<HTMLSelectElement>,
-        HTMLSelectElement
-    >;
-    children: ReactNode;
+    nativeSelectProps?: Omit<
+        React.DetailedHTMLProps<React.SelectHTMLAttributes<HTMLSelectElement>, HTMLSelectElement>,
+        "value" | "defaultValue"
+    > & {
+        // Overriding the type of value and defaultValue to only accept the value type of the options
+        value?: Options[number]["value"];
+        defaultValue?: Options[number]["value"];
+    };
     /** Default: false */
     disabled?: boolean;
     /** Default: "default" */
     state?: "success" | "error" | "default";
     /** The message won't be displayed if state is "default" */
     stateRelatedMessage?: ReactNode;
     style?: CSSProperties;
+    placeholder?: string;
 };
+export type GenericOption<OptionValue> = {
+    value: OptionValue;
+    label: string;
+    disabled?: boolean;
+    hidden?: boolean;
+    selected?: boolean;
+};
+
+type DefaultOptionValue = string | number | readonly string[] | undefined;
 
 /**
  * @see <https://react-dsfr-components.etalab.studio/?path=/docs/components-select>
  * */
-export const Select = memo(
-    forwardRef<HTMLDivElement, SelectProps>((props, ref) => {
-        const {
-            className,
-            label,
-            hint,
-            nativeSelectProps,
-            disabled = false,
-            children,
-            state = "default",
-            stateRelatedMessage,
-            style,
-            ...rest
-        } = props;
+const SelectComponent = <T extends GenericOption<DefaultOptionValue>[]>(
+    props: SelectProps<T>,
+    ref: React.LegacyRef<HTMLDivElement>
+) => {
+    const {
+        className,
+        label,
+        hint,
+        nativeSelectProps,
+        disabled = false,
+        options,
+        state = "default",
+        stateRelatedMessage,
+        placeholder,
+        style,
+        ...rest
+    } = props;
 
-        assert<Equals<keyof typeof rest, never>>();
+    assert<Equals<keyof typeof rest, never>>();
+    const selectIdExplicitelyProvided = nativeSelectProps?.id;
```

Comment:
This should be declared within the closure, it dosen't need to have a wider scope.